### PR TITLE
[TEP-0117] Tekton Results Logs

### PR DESCRIPTION
**Pull request by sayan-biswas**
# Description

[[TEP-0117] Tekton Results Logs](https://github.com/tektoncd/community/blob/main/teps/0117-tekton-results-logs.md)

This PR extends [Tekton Results](https://github.com/tektoncd/results) to support log persistence outside of the host Kubernetes cluster. With this feature results will be able to store `PipelineRun` and `TaskRun` logs in the following persistent locations:

- File (PVC)
- S3 (S3 compatible storage)

# Release Notes
```release-note
Kubernetes version: 1.22.x
Tekton Pipeline version: 0.42.0

Add support for storing PipelineRun and TaskRun logs in persistent storage backends.
Refer to the TEP-0117 for more details.
```

### CODE REVIEW

**Review by khrm — COMMENTED** [review]
/ok-to-test

**Review by adambkaplan — COMMENTED** [review]
Not sure how other folks feel - I prefer to have vendor updates isolated in a separate commit. That makes it easier to a) review the code that actually changed, and b) rebase if there are conflicts in go.mod.

**Comment by sayan-biswas** [comment]
> Not sure how other folks feel - I prefer to have vendor updates isolated in a separate commit. That makes it easier to a) review the code that actually changed, and b) rebase if there are conflicts in go.mod.

Sure, I'll separate this in two commits.

**Comment by khrm** [comment]
We need some documentation also.

I tried testing this in OpenShift using `File(PVC)`. Somehow Logs aren't coming. Will try again on Monday.
https://github.com/sayan-biswas/tekton-results/compare/logs-server...khrm:results:logs-server?expand=1

File: pkg/logs/writer_test.go [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+	}
+	_, err := m.receivedData.Write(log.GetData())
+	return err
+}
+
+// Test case: "Hello!" has got 6 bytes. Chunks size is 5. Remain is 6 - 5 = 1.
+// NewBufferedWriter.Write([]byte) should send 5 bytes(1 chunk) to sender
+// and save 1 byte in the buffer.
+// NewBufferedWriter.WriteRemain() should send last byte to sender.
+func TestBufferedWriterWhenContentIsLongerThenChunk(t *testing.T) {
```
Comment (adambkaplan) [inline pkg/logs/writer_test.go:26]: Consider making these unit tests [table-driven](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests), since the validations are similar.

File: pkg/api/server/v1alpha2/logs.go [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+		s.logger.Errorf("no logs exist for %s", req.GetName())
+		return status.Error(codes.NotFound, "Log doesn't exist")
+	}
+
+	writer := logs.NewBufferedWriter(srv, name, s.config.LOGS_CHUNK_SIZE)
+	if _, err = streamer.WriteTo(writer); err != nil {
+		s.logger.Error(err)
+		return status.Error(codes.Internal, "Error streaming log")
+	}
+	_, err = writer.WriteRemain()
```
Comment (adambkaplan) [inline pkg/api/server/v1alpha2/logs.go:70]: See my suggestion to rename this function `Flush` - debatable if that makes intent clearer. No matter what, however, we should godoc that this function must be called after writes are completed.

**Review by adambkaplan — COMMENTED** [review]
/approve

Marking approve to unblock other reviewers - I have outlined a few items that are mostly stylistic recommendations than things to review wrt functionality. There are a few asks with respect to unit tests as well.

File: pkg/api/server/v1alpha2/log/s3.go [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+	DefaultS3MultiPartSize = 1024 * 1024 * 5
+)
+
+type s3Client interface {
+	AbortMultipartUpload(context.Context, *s3.AbortMultipartUploadInput, ...func(*s3.Options)) (*s3.AbortMultipartUploadOutput, error)
+	CompleteMultipartUpload(context.Context, *s3.CompleteMultipartUploadInput, ...func(*s3.Options)) (*s3.CompleteMultipartUploadOutput, error)
+	CreateMultipartUpload(context.Context, *s3.CreateMultipartUploadInput, ...func(*s3.Options)) (*s3.CreateMultipartUploadOutput, error)
+	DeleteObject(context.Context, *s3.DeleteObjectInput, ...func(*s3.Options)) (*s3.DeleteObjectOutput, error)
+	GetObject(context.Context, *s3.GetObjectInput, ...func(*s3.Options)) (*s3.GetObjectOutput, error)
+	UploadPart(context.Context, *s3.UploadPartInput, ...func(*s3.Options)) (*s3.UploadPartOutput, error)
```
Comment (adambkaplan) [inline pkg/api/server/v1alpha2/log/s3.go:32]: This is the right pattern - the new AWS SDK recommends using your own interfaces to mock/stub in unit tests:

https://aws.github.io/aws-sdk-go-v2/docs/unit-testing/

Comment (sayan-biswas) [reply pkg/api/server/v1alpha2/log/s3.go:31]: Added unit tests covering basic cases for S3

File: pkg/logs/writer_test.go (hunk 1 above)
Comment (sayan-biswas) [reply pkg/logs/writer_test.go]: Changed to table to driven.

File: docs/DEVELOPMENT.md [hunk 4]
```diff
@@ -84,9 +84,8 @@ tekton.results.v1alpha2.Results
 
 # Create gRPC access token
 $ export GRPC_ACCESS_TOKEN=$(kubectl create token tekton-results-debug -n tekton-pipelines)
-
 # Makes a request to the Results service.
-$ grpc_cli call --channel_creds_type=ssl --ssl_target=tekton-results-api-service.tekton-pipelines.svc.cluster.local --call_creds=access_token=$GRPC_ACCESS_TOKEN localhost:50051 tekton.results.v1alpha2.Results.ListResults 'parent: "default"'
+$ grpc_cli call --channel_creds_type=ssl --ssl_target=tekton-results-api-service.tekton-pipelines.svc.cluster.local --call_creds=access_token=$GRPC_ACCESS_TOKEN localhost:50051 tekton.results.v1alpha2.Results.ListResults 'parent: "default"'grpc_cli call --channel_creds_type=ssl --ssl_target=tekton-results-api-service.tekton-pipelines.svc.cluster.local --call_creds=access_token=$(kubectl get secrets -n tekton-pipelines -o jsonpath="{.items[?(@.metadata.annotations['kubernetes\.io/service-account\.name']=='tekton-results-debug')].data.token}"|base64 --decode) localhost:50051 tekton.results.v1alpha2.Results.ListResults 'parent: "default"'
```
Comment (avinal) [inline docs/DEVELOPMENT.md:88]:
```suggestion
$ grpc_cli call --channel_creds_type=ssl --ssl_target=tekton-results-api-service.tekton-pipelines.svc.cluster.local --call_creds=access_token=$GRPC_ACCESS_TOKEN localhost:50051 tekton.results.v1alpha2.Results.ListResults 'parent: "default"
```

This should probably be removed.

Comment (adambkaplan) [reply docs/DEVELOPMENT.md:88]: good catch - looks like some of the grpc calls were copied by accident.

**Comment by adambkaplan** [comment]
@dibyom are you able to take a look at this PR soon?

@alan-ghelardi you may also be interested.

**Comment by dibyom** [comment]
Yes, looking today

File: cmd/api/README.md [hunk 5]
```diff
@@ -14,9 +14,19 @@
...
+| S3_BUCKET_NAME        | S3 Bucket name                                                                                                           | elasticbeanstalk-ap-south-1-186765682378     |
+| S3_ENDPOINT           | S3 Endpoint                                                                                                              | https://s3.ap-south-1.amazonaws.com          |
+| S3_HOSTNAME_IMMUTABLE | S3 Hostname immutable                                                                                                    | false (default)                              |
+| S3_REGION             | S3 Region                                                                                                                | ap-south-1                                   |
+| S3_ACCESS_KEY_ID      | S3 Access Key ID                                                                                                         | AKIGFT37RLYHUJ4GRE7M                         |
+| S3_SECRET_ACCESS_KEY  | S3 Secret Access Key                                                                                                     | +9btkSFv42sTW+Kl/HYhzmUBM+gP2y9kzJ354hrJ     |
```
Comment (dibyom) [inline cmd/api/README.md:28]: Umm...just making sure- are these actual secret key/access keys? (Have they been revoked/rotated?)

Comment (adambkaplan) [reply cmd/api/README.md:28]: @sayan-biswas relayed to me that these are not real (they are missing a few characters and are thus not valid). We can follow up and make this more generic 

```suggestion
| S3_SECRET_ACCESS_KEY  | S3 Secret Access Key                                                                                                     | <your AWS_SECRET_ACCESS_KEY>    |
```

**Comment by adambkaplan** [comment]
/hold

Current downstream testing with minio is failing (this is something we can consider for upstream testing of S3, too).

**Comment by sayan-biswas** [comment]
Fixed (minio compatible changes)

Tests:
image


Mino:
image

File: cmd/api/README.md (hunk 5 above)
Comment (sayan-biswas) [reply cmd/api/README.md]: updated with generic text.

File: pkg/api/server/v1alpha2/log/s3.go [hunk 6]
```diff
@@ -155,38 +157,39 @@ func (s3s *s3Stream) ReadFrom(r io.Reader) (int64, error) {
 		return 0, err
 	}
 
-	size := s3s.partSize + n
+	size := s3s.totalSize + n
```
Comment (dibyom) [inline pkg/api/server/v1alpha2/log/s3.go:160]: question: why was this renamed from partSize to totalSize?

File: pkg/watcher/convert/convert.go [hunk 7]
```diff
@@ -47,6 +51,40 @@ func ToProto(in runtime.Object) (*rpb.Any, error) {
 	}, nil
 }
 
+func ToLogProto(in metav1.Object, kind, name string) (*rpb.Any, error) {
```
Comment (dibyom) [inline pkg/watcher/convert/convert.go:54]: Can we add a unit test for this?

File: pkg/watcher/reconciler/dynamic/dynamic_test.go [hunk 8]
```diff
@@ -251,7 +252,7 @@ func TestReconcile_PipelineRun(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	r := NewDynamicReconciler(results, prclient, nil)
+	r := NewDynamicReconciler(resultsClient, logsClient, prclient, nil)
```
Comment (dibyom) [inline pkg/watcher/reconciler/dynamic/dynamic_test.go:255]: Are we doing any testing with the logsClient in these tests to see if the log file was actually written?

File: go.mod [hunk 9]
```diff
@@ -3,104 +3,111 @@ module github.com/tektoncd/results
...
-	github.com/jackc/pgconn v1.11.0
+	github.com/grpc-ecosystem/grpc-gateway/v2 v2.15.0
+	github.com/jackc/pgconn v1.13.0
 	github.com/jonboulle/clockwork v0.3.0
 	github.com/mattn/go-sqlite3 v2.0.3+incompatible
 	github.com/prometheus/client_golang v1.14.0
 	github.com/spf13/viper v1.14.0
-	github.com/tektoncd/pipeline v0.39.0
-	go.uber.org/automaxprocs v1.5.1
+	github.com/tektoncd/pipeline v0.42.0
```
[the remark this answers is not in the document]
Comment (dibyom) [reply go.mod:24]: We should mention in the PR description/release notes section the minimum version of k8s required to run results now. (As well as the version of tekton pipelines supported)

File: pkg/logs/writer.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+	}
+	return &BufferedLog{
+		sender:    sender,
+		name:      name,
+		chunkSize: chunkSize,
+		buffer:    *bytes.NewBuffer(make([]byte, 0)),
+	}
+}
+
+func (w *BufferedLog) Write(p []byte) (n int, err error) {
```
[the remark this answers is not in the document]
Comment (dibyom) [reply pkg/logs/writer.go:42]: +1

File: pkg/api/server/v1alpha2/log/file.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+	}
+}
+
+func (*fileStream) Type() string {
+	return string(v1alpha2.FileLogType)
+}
+
+// WriteTo reads the contents of the TaskRun log file and writes them to the provided writer, such
+// as os.Stdout.
+func (fs *fileStream) WriteTo(w io.Writer) (n int64, err error) {
```
[the remark this answers is not in the document]
Comment (sayan-biswas) [reply pkg/api/server/v1alpha2/log/file.go:49]: unit test added.

File: pkg/api/server/v1alpha2/log/file.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+	}()
+	// Use the bufferred reader to ensure file contents are not read entirely into memory
+	reader := bufio.NewReaderSize(file, fs.bufSize)
+	n, err = reader.WriteTo(w)
+	return
+}
+
+// ReadFrom reads the log contents from the provided io.Reader, and writes them to the TaskRun log
+// file on disk.
+func (fs *fileStream) ReadFrom(r io.Reader) (n int64, err error) {
```
[the remark this answers is not in the document]
Comment (sayan-biswas) [reply pkg/api/server/v1alpha2/log/file.go:72]: unit test added

**Comment by dibyom** [comment]
Hi @sayan-biswas thanks for the PR and sorry for the delay in reviewing. A couple of housekeeping items but overall looks good
1. Could you add the min version of k8s and the version of tekton pipelines supported with the updates to the k8s and pipeline dependencies?

2. It would be useful to have a bit more [detail in the commit messages](https://github.com/tektoncd/community/blob/main/standards.md#commits) - at the very least we should reference the TEP number in the PR description/commit message so in the future we have a bit more context on the design

/approve

File: pkg/api/server/v1alpha2/log/s3.go (hunk 6 above)
Comment (sayan-biswas) [reply pkg/api/server/v1alpha2/log/s3.go:160]: Nothing in particular, "totalSize" seemed more relevant name, and for Minio tenant the ContentLength was required in the API input params later in the method uploadPart, which is named partSized. Thought this would avoid confusion.

File: pkg/watcher/reconciler/dynamic/dynamic_test.go (hunk 8 above)
Comment (sayan-biswas) [reply pkg/watcher/reconciler/dynamic/dynamic_test.go:255]: The API handler for this has a unit test, but from reconciling -> API handler -> logs written doesn't have a test. As of now we cannot add this kind of test S3 backend, but test for File storage backend can be implemented.

File: pkg/watcher/convert/convert.go (hunk 7 above)
Comment (sayan-biswas) [reply pkg/watcher/convert/convert.go:54]: Sure, I'll add it

**Comment by khrm** [comment]
@avinal Did you test using minio? Or File storage? It's failing when I test file storage.

image

**Comment by avinal** [comment]
@khrm I used file storage for testing.

**Comment by avinal** [comment]
I tried testing this on an OpenShift cluster on @khrm suggestion. there seem to be many issues, gRPC calls fail for logs giving SSL_ERROR but all other methods work fine. With curl, there is no connection issue but it shows no logs. One problem I found is the status is showing "unknown" for all PipelineRuns and TaskRuns. it makes sense if PR/TR is not complete, there cannot be any logs. I tried this two times, same errors. 

Is there something I am missing? I followed the same procedure as Kind cluster to install on OpenShift Cluster.

**Comment by sayan-biswas** [comment]
@khrm Did you mount a PVC and set the path here in the config below?
https://github.com/tektoncd/results/blob/78b4475507a4ea00bc5c73a21811e1db6a03e778/config/env/config#L17

**Comment by khrm** [comment]
@sayan-biswas 
In addition to adding PVC, I had to add the following rules to `watcher` Clusterrole
https://github.com/khrm/results/blob/f879625aacbd55b6e7ea3237d03d95c0381df279/config/100-watcher-serviceaccount.yaml#L43-L48


``` 
  - apiGroups: [""]
    resources: ["pods"]
    verbs: ["get", "list", "watch"]
  - apiGroups: [""]
    resources: ["pods/log"]
    verbs: ["get"]

```
This contains the every changes that I did to make logging (File storage) work in OpenShift:
https://github.com/sayan-biswas/tekton-results/compare/logs-server...khrm:results:logs-server?expand=1
image


/approve

We can proceed with merging this pr after required changes mentioned by @dibyom . Probably, you also need to add those rules to watcher serviceaccount.

File: pkg/api/server/v1alpha2/log/s3.go (hunk 6 above)
Comment (sayan-biswas) [reply pkg/api/server/v1alpha2/log/s3.go]: Reverted back.

File: pkg/watcher/convert/convert.go (hunk 7 above)
Comment (sayan-biswas) [reply pkg/watcher/convert/convert.go:54]: Add unit test for this.

File: pkg/watcher/reconciler/dynamic/dynamic_test.go (hunk 8 above)
Comment (sayan-biswas) [reply pkg/watcher/reconciler/dynamic/dynamic_test.go:291]: @adambkaplan @dibyom 
The problem in implementing this in the reconciler unit test is that we are using `tkncli` to collect logs, which is getting it from the `pod` logs. But in these unit tests we are creating only `PipelineRuns` and `TaskRuns` but no pods.
We have covered writing of the logs in the `UpdateLog` handler of the API server, and in this case we can use an e2e test, which will be easier.

**Comment by AndrienkoAleksandr** [comment]
pr looks good, but I think  `S3_ACCESS_KEY_ID` and `S3_SECRET_ACCESS_KEY` are security sensitive info, so we should provide them using secret instead of configmap. I think secret can be optional in the deployment.

**Comment by sayan-biswas** [comment]
@AndrienkoAleksandr The `viper.AutomaticEnv()` will automatically load data from any environment variables with the same name in the `struct config {}` and will have higher priority than loading from config file. Then we can use the secrets to load those environment variables. 
So code change will not be required.

**Comment by adambkaplan** [comment]
@dibyom @alan-ghelardi I think it's time to make a "last call for reviews" on this PR. I think most of the items raised have been addressed.

**Comment by sayan-biswas** [comment]
@dibyom Some changes were required to rebase with the PR #334, hence the LGTM is removed.
@alan-ghelardi My PR is already passing the `config` object to Server's `New` function. So I have added the **AutoMigrate** condition directly from the `config` object.

https://github.com/tektoncd/results/blob/db2ede53ea3e30e2be7cd7a11749cbc0728023be/pkg/api/server/v1alpha2/server.go#L84-L88

**Review by khrm — COMMENTED** [review]
/kind feature
We need this label for release as discussed in the last WG call.
